src/utils.ts

```diff
@@ -0,0 +1,564 @@
+ */
+export const recursivelyDownloadTx = async (blockId: string, txIds: string[] = [], data: FullTx[] = []): Promise<FullTx[]> => {
+  if (txIds.length === 0) {
+    return data;
+  }
+
+  const txId = txIds.pop();
+  const txData = await downloadTx(txId);
+  const { tx, meta } = txData;
+  const parsedTx: FullTx = parseTx(tx);
+
+  if (parsedTx.parents.length > 2) {
+    // We downloaded a block, we should ignore it
+    return recursivelyDownloadTx(blockId, txIds, data);
+  }
+
+  // If the first_block from the downloaded tx is not from the block we are searching, we should ignore it.
+  if (meta.first_block !== blockId) {
+    return recursivelyDownloadTx(blockId, txIds, data);
+  }
+
+  // Ignore voided txs
+  if (meta.voided_by && meta.voided_by.length && meta.voided_by.length > 0) {
+    return recursivelyDownloadTx(blockId, txIds, data);
+  }
+
+  const newParents = parsedTx.parents.filter((parent) => txIds.indexOf(parent) < 0 && parent !== txId);
+
+  return recursivelyDownloadTx(blockId, [...txIds, ...newParents], [...data, parsedTx]);
+};
+
+/**
+ * Prepares a transaction to be sent to the wallet-service `onNewTxRequest`
+ *
+ * @param tx - `FullTx` or `FullBlock` representing a typed transaction to be prepared
+ */
+export const prepareTx = (tx: FullTx | FullBlock): PreparedTx => {
+  const prepared = {
+    ...tx,
+    tx_id: tx.txId,
+    raw: '',
+    inputs: tx.inputs.map((input) => {
+      const baseInput: PreparedInput = {
+        value: input.value,
+        token_data: input.tokenData,
+        script: input.script,
+        token: input.token,
+        decoded: input.decoded,
+        index: input.index,
+      };
+
+      if (input.tokenData === 0) {
+        return {
+          ...baseInput,
+          token: '00',
+        };
+      }
+
+      /* eslint-disable */
+      const { uid } = tx.tokens[(input.tokenData & TOKEN_INDEX_MASK) - 1]; // eslint-disable-line no-bitwise
+      /* eslint-enable */
+
+      return {
+        ...baseInput,
+        token: uid,
+      };
+    }),
+    outputs: tx.outputs.map((output) => {
+      const baseOutput: PreparedOutput = {
+        value: output.value,
+        token_data: output.tokenData,
+        script: output.script,
+        token: output.token,
+        spent_by: output.spentBy,
+        decoded: output.decoded,
+      };
+
+      if (output.tokenData === 0) {
+        return {
+          ...baseOutput,
+          token: '00',
+        };
+      }
+
+      /* eslint-disable */
+      const { uid } = tx.tokens[(output.tokenData & TOKEN_INDEX_MASK) - 1]; // eslint-disable-line no-bitwise
+      /* eslint-enable */
+
+      return {
+        ...baseOutput,
+        token: uid,
+      };
+    }),
+  };
+
+  return prepared;
+};
+
+/**
+ * Types a tx that was received from the full_node
+ *
+ * @param tx - The transaction object as received by the full_node
+ */
+export const parseTx = (tx: any): FullTx => {
```

Comment:
I feel we should move it to a `utils.js` file.

Reply:
Done in c5e1e4b4b2b7be18707e3e1c61f18e5bd758fe49